### PR TITLE
Smooch type definitions

### PR DESCRIPTION
**Pull request by jpsorensen**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by peterblazejewicz** [comment]
@jpsorensen is this ESM only module or should it work CJS module (with support for 'default' via es interop flags?
https://github.com/zendesk/sunshine-conversations-web#browserify-and-webpack

**Comment by peterblazejewicz** [comment]
and congratulations on your first PR :)

**Comment by jpsorensen** [comment]
Thanks @peterblazejewicz! You're right, this should work as a global script and via require(). I'm honestly still wrapping my head around how to correctly export this for the various module formats so I'm going to take some time to try to understand this better and hopefully get it right. Appreciate your patience!

> what's the audience here?

Apologies for my ignorance; I think you're asking how the library is typically used? It would be used in client-side web applications (it's essentially a chat widget).

**Comment by peterblazejewicz** [comment]
>  It would be used in client-side web applications (it's essentially a chat widget).

So assume the must have is this support for ESM/CJS exports. The global script would be welcome. Let me know if you need a hand with this one. 
Technically this is not UMD library:
https://www.typescriptlang.org/docs/handbook/modules.html#umd-modules
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-a-umd-library
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#examples-of-umd-libraries
So this is IMO up to you to decide how to structure those types. First correct this export to match CJS  module, it will work in ESM modules without a problem.

**Comment by peterblazejewicz** [comment]
@jpsorensen here is sample of one of approach to achieve both, if that would help:
https://gist.github.com/peterblazejewicz/f5d5436414865626078f5a2d16f78a69
you can give it a try

**Comment by jpsorensen** [comment]
Thanks for the pointers and help - I hope you don't mind I've adopted exactly the export style you recommended (and updated the unit tests to use CJS-style require() import). Thanks so much for your help!

File: types/smooch/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,954 @@
...
+        beforePostbackSend?: (postback: unknown, data: ConversationData) => unknown;
+        onInvalidAuth?: () => string | Promise<string>;
+    }
+}
+
+type Nullable<T> = { [P in keyof T]: T[P] | null };
+
+declare const smooch: typeof Smooch;
+
+export as namespace Smooch;
```
Comment (andrewbranch) [inline types/smooch/index.d.ts:953]: Looking at the JS on unpkg, I don’t actually see a UMD wrapper—it looks like just a CJS `module.exports`, which would mean you can’t use as a global just by including a script tag and you should drop this line.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: types/smooch/index.d.ts (hunk 1 above)
Comment (peterblazejewicz) [reply types/smooch/index.d.ts:953]: > Looking at the JS on unpkg, I don’t actually see a UMD wrapper

yes, this is what we wrote, this is not an UMD wrapper. We just discussed how to approach a situation, if someone writes a non-module code, adding a script tag to the page, which still can find DT definition useful, say in VSCode:
The 'Smooch' object, (net yet unitialized) become available in global script:
```
Smooch
{init: ƒ, on: ƒ, render: ƒ, destroy: ƒ}
```

**Review by andrewbranch — APPROVED** [review]
Oops, I missed some of the earlier comments discussing script tag usage. Unusual but I think this is best given the README lists global usage. Thanks!

**Comment by jpsorensen** [comment]
Ready to merge

**Comment by peterblazejewicz** [comment]
@jpsorensen now congrats on first merged, nice!

**Comment by jpsorensen** [comment]
Thank you, excited to be contributing to the Typescript ecosystem. Really appreciate your help and your time!

**Pull request merged**